legate/core/_lib/types.pyx

```diff
@@ -0,0 +1,293 @@
+        _SUB "SUB_LT"
+        _MUL "MUL_LT"
+        _DIV "DIV_LT"
+        _MAX "MAX_LT"
+        _MIN "MIN_LT"
+        _OR  "OR_LT"
+        _AND "AND_LT"
+        _XOR "XOR_LT"
+
+BOOL        = legate_core_type_code_t._BOOL
+INT8        = legate_core_type_code_t._INT8
+INT16       = legate_core_type_code_t._INT16
+INT32       = legate_core_type_code_t._INT32
+INT64       = legate_core_type_code_t._INT64
+UINT8       = legate_core_type_code_t._UINT8
+UINT16      = legate_core_type_code_t._UINT16
+UINT32      = legate_core_type_code_t._UINT32
+UINT64      = legate_core_type_code_t._UINT64
+FLOAT16     = legate_core_type_code_t._FLOAT16
+FLOAT32     = legate_core_type_code_t._FLOAT32
+FLOAT64     = legate_core_type_code_t._FLOAT64
+COMPLEX64   = legate_core_type_code_t._COMPLEX64
+COMPLEX128  = legate_core_type_code_t._COMPLEX128
+FIXED_ARRAY = legate_core_type_code_t._FIXED_ARRAY
+STRUCT      = legate_core_type_code_t._STRUCT
+STRING      = legate_core_type_code_t._STRING
+INVALID     = legate_core_type_code_t._INVALID
+
+ADD = legate_core_reduction_op_kind_t._ADD
+SUB = legate_core_reduction_op_kind_t._SUB
+MUL = legate_core_reduction_op_kind_t._MUL
+DIV = legate_core_reduction_op_kind_t._DIV
+MAX = legate_core_reduction_op_kind_t._MAX
+MIN = legate_core_reduction_op_kind_t._MIN
+OR  = legate_core_reduction_op_kind_t._OR
+AND = legate_core_reduction_op_kind_t._AND
+XOR = legate_core_reduction_op_kind_t._XOR
+
+_NUMPY_DTYPES = {
+    BOOL : np.dtype(np.bool_),
+    INT8 : np.dtype(np.int8),
+    INT16 : np.dtype(np.int16),
+    INT32 : np.dtype(np.int32),
+    INT64 : np.dtype(np.int64),
+    UINT8 : np.dtype(np.uint8),
+    UINT16 : np.dtype(np.uint16),
+    UINT32 : np.dtype(np.uint32),
+    UINT64 : np.dtype(np.uint64),
+    FLOAT16 : np.dtype(np.float16),
+    FLOAT32 : np.dtype(np.float32),
+    FLOAT64 : np.dtype(np.float64),
+    COMPLEX64 : np.dtype(np.complex64),
+    COMPLEX128 : np.dtype(np.complex128),
+    STRING : np.dtype(np.str_),
+}
+
+
+cdef extern from "core/type/type_info.h" namespace "legate" nogil:
+    cdef cppclass Type:
+        ctypedef enum Code:
+            pass
+        int code
+        unsigned int size()
+        unsigned int alignment()
+        int uid()
+        bool variable_size()
+        unique_ptr[Type] clone()
+        string to_string()
+        int find_reduction_operator(int) except+
+
+    cdef cppclass FixedArrayType(Type):
+        unsigned int num_elements()
+        const Type& element_type()
+
+    cdef cppclass StructType(Type):
+        unsigned int num_fields()
+        const Type& field_type(unsigned int)
+        bool aligned()
+
+    cdef unique_ptr[Type] primitive_type(int code)
+
+    cdef unique_ptr[Type] string_type()
+
+    cdef unique_ptr[Type] fixed_array_type(
+        unique_ptr[Type] element_type, unsigned int N
+    ) except+
+
+    cdef unique_ptr[Type] struct_type_raw_ptrs(
+        vector[Type*] field_types, bool
+    ) except+
+
+
+cdef Dtype from_ptr(Type* ty):
```

Comment:
Would it be possible to use forwarding references here, similar to https://github.com/cython/cython/blob/master/tests/run/cpp_forwarding_ref.pyx, to avoid having to `release()` then `reset()` the `unique_ptr`s?